store/data_source.go

```diff
@@ -536,3 +553,200 @@ func (*Store) listDataSourceV2(ctx context.Context, tx *Tx, instanceID string) (
 
 	return dataSourceMessages, nil
 }
+
+// AddDataSourceToInstanceV2 adds a RO data source to an instance and return the instance where the data source is added.
+func (s *Store) AddDataSourceToInstanceV2(ctx context.Context, instanceUID, creatorID int, dataSource *DataSourceMessage) (*InstanceMessage, error) {
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return nil, errors.New("Failed to begin transaction")
+	}
+	defer tx.Rollback()
+
+	// We flatten the data source fields in DataSourceMessage, so we need to compose them in store layer before INSERT.
+	dataSourceOptions := api.DataSourceOptions{
+		SRV:                    dataSource.SRV,
+		AuthenticationDatabase: dataSource.AuthenticationDatabase,
+	}
+
+	if _, err := tx.QueryContext(ctx, `
+		INSERT INTO data_source (
+			creator_id,
+			updater_id,
+			instance_id,
+			database_id,
+			name,
+			type,
+			username,
+			password,
+			ssl_key,
+			ssl_cert,
+			ssl_ca,
+			host,
+			port,
+			options,
+			database
+		)
+		SELECT $1, $2, $3, data_source.database_id, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14
+			FROM data_source WHERE instance_id = $15 AND type = $16;
+	`, creatorID, creatorID, instanceUID,
+		dataSource.Title, dataSource.Type, dataSource.Username,
+		dataSource.Password, dataSource.SslKey, dataSource.SslCert,
+		dataSource.SslCa, dataSource.Host, dataSource.Port,
+		dataSourceOptions, dataSource.Database, instanceUID, api.Admin,
```

Comment:
If we want to merge this with CreateInstanceV2, we cannot use query with ADMIN type, but SELECT from database with name == AllDatabaseName ("*").

Reply:
Done.
It's ok for me now. But we need to find another way to link the data source to other databases in the future.